### ISSUE TITLE
Make ComboBox accessible

### ISSUE THREAD

**Issue by jessegreenberg** [phetsims/sun#314]
From https://github.com/phetsims/molarity/issues/37, we would like to make ComboBox accessible with keyboard navigation and description. We found recommendation from W3 for best practices and a working example:

https://www.w3.org/TR/2016/WD-wai-aria-practices-1.1-20160317/#combobox
http://demos.telerik.com/kendo-ui/combobox/keyboard-navigation

but it would be great to have a more solid mockup for the accessible HTML before we start adding to ComboBox. @terracoda is going to work on this and provide an example.  Assigning to @terracoda, could you please paste links in this issue and reassign to me when ready?

**Comment by terracoda** [comment]
Adding some resources for review:
- [https://www.levelaccess.com/differences-aria-1-0-1-1-changes-rolecombobox/](https://www.levelaccess.com/differences-aria-1-0-1-1-changes-rolecombobox/)

- [https://labs.ssbbartgroup.com/index.php/ARIA_Combobox](https://labs.ssbbartgroup.com/index.php/ARIA_Combobox)

- [https://www.w3.org/blog/wai-components-gallery/widget/combobox-with-aria-autocompleteinline/](https://www.w3.org/blog/wai-components-gallery/widget/combobox-with-aria-autocompleteinline/) maybe not relevant as the text is not editable in the sim?

**Comment by terracoda** [comment]
My understanding is that work on this issue is happening in https://github.com/phetsims/a11y-research/issues/64, and we are almost there with an accessible combobox.

**Comment by terracoda** [comment]
@jessegreenberg, I have made further recommendations in https://github.com/phetsims/a11y-research/issues/64#issuecomment-364152234

and am re-assigning these 2 issues for you to distribute.

**Comment by zepumph** [comment]
From meetings today, @emily-phet asked me to work on ComboBox for Molarity. My priorities align nicely to take a look at this in the next week or so.

**Comment by jessegreenberg** [comment]
@terracoda pointed out in #363 design of the general ComboBox could impact this issue. I am not sure if a grid-like combobox was considered in our initial investigation, or whether a 2D combobox will be a totally different a11y implementation.

**Comment by zepumph** [comment]
Using design from https://github.com/phetsims/a11y-research/issues/64. @emily-phet pointed out this morning that I should just implement the current structure. The first use case will be for Molarity.

**Comment by terracoda** [comment]
@zepumph, yes please make a general combobox for Molarity. I will investigate those other patterns as soon as possible, but not immediately.

**Comment by zepumph** [comment]
First step was to make ComboBox es6.

**Comment by zepumph** [comment]
@jessegreenberg pointed me to an old branch that had one implementation of it. I'm going to see how far putting that on master takes me, with the goal of eventually supporting the html in `ComboBox.md`:

```html
	<div tabindex="-1" id="container-for-labels">
	<span id="listbox-static-label">Solute</span>
	  <button id="listbox-option-dynamic-label" tabindex="0" aria-haspopup="listbox" aria-labelledby="listbox-static-label listbox-option-dynamic-label">Drink Mix</button>
	</div>
	<ul role="listbox" tabindex="0" id="listbox" aria-activedescendant="option-1" aria-labelledby="listbox-static-label" style="list-style:none;">
	  <li role="option" id="option-1" class="selected" aria-selected="true">Drink mix</li>
	  <li role="option" id="option-2">Cobalt (II) nitrate</li>
	  <li role="option" id="option-3">Cobalt Chloride</li>
	  <li role="option" id="option-4">Potassium dichromate</li>
	  <li role="option" id="option-5">Gold (III) chloride</li>
	  <li role="option" id="option-6">Potassium chromate</li>
	  <li role="option" id="option-7">Nickel (II) chloride</li>
	  <li role="option" id="option-8">Copper sulfate</li>
	  <li role="option" id="option-9">Potassium permanganate</li>
	  <li role="option" id="option-10">Potassium dichromate</li>
	</ul>
	<!-- help text for combobox -->
	<p>Change a solute and observe differences.</p>
```

**Comment by terracoda** [comment]
@zepumph, if it is helpful there was an example in the a11y-research  repo in the html sketches folder as well https://github.com/phetsims/a11y-research/blob/master/html-sketches/combobox/combobox.html. 

This might be the same HTML as what @jessegreenberg posted above and what's in the combobox.md file.

There are several open issues about combobox, so I am having a little trouble weaving them together.

**Comment by zepumph** [comment]
Thanks @terracoda. I will use that link to combobox.html to base my current implementation on.

For now I am completely ignoring the intersection of mouse and a11y. I think it will be simple enough to support that once we have solidified our direction with the PDOM.

**Comment by zepumph** [comment]
After the above commit, everything is implemented except active descendant. I will need to implement it first in https://github.com/phetsims/scenery/issues/873. I'll try to get it done this afternoon, and will check back here afterwards with an update.

**Comment by zepumph** [comment]
Added aria-activedescendant support above. I am ready for combobox to be reviewed. 

@terracoda, @jessegreenberg, and @emily-phet please have a listen here:
https://phet-dev.colorado.edu/html/molarity/1.5.0-dev.1/phet/molarity_en_phet.html?a11y

I took a brief listen on NVDA chrome and it sounds nice! Note that there aren't other labels on things, so just worry about the sound of the comboBox.

You can also see in the a11y view that the HTML is the same as https://github.com/phetsims/a11y-research/blob/master/html-sketches/combobox/combobox.html. 

I still have a few todos in ComboBox.js, but I thought it would be good to get an idea about people's first impressions.

My checklist of still to do's:
- [ ] Multiple todos pointing to this issue.
- [ ] Use of aria-desribedby that I removed? (see https://github.com/phetsims/sun/issues/314#issuecomment-433653144)
- [ ] Implement mouse/a11y interface, partially relating to https://github.com/phetsims/scenery/issues/58 since there is a workaround in ComboBox for that which complicates things for us.

Note to @pixelzoom that molarity was only committed on to add the `a11yLabel` option. I was surprised at how non-invasive it was, though we will likely rename that option at some point.

**Comment by terracoda** [comment]
@zepumph, this sounds really good in VoiceOver with the exception that VoiceOver adds the word "text" after the name of each option. I think this is because the role option assumes a text input. I don't think we can do anything about that.

VoiceOver announces both labels, "Solute" and the selected solute upon focus. So on focus with VoiceOver, I hear:

> Solute Drink Mix, popup button 
which is nice!

And when I open the list (or listbox) of options, I hear the name of the selected item and list's count semantics:
> Drink Mix, text (1 of 9)

I also tested in Chrome with ChromeVox. Unfortuantely, ChromeVox does not announce both labels on focus, but the selected item and listbox are both very clear once the listbox is popped up.

The pink focus highlights are all fine, though we may want to consider a dotted outline (or a light pink background color similar to the light blue shown for mouse hover) to show focus moving within the listbox before a new selection is made. The focus highlight moves which is good, but we may want to communicate what is actually selected when the two are not the same.

Nice work @zepumph!

**Comment by terracoda** [comment]
@zepumph, I just noticed that the keyboard interaction is not working in Firefox Quantum 62.0.3 on MacOS High Sierra 10.13.6

When I try to use the arrow keys on the popped up list the focus highlight gets lost.  This was using Firefox without a screen reader.

**Comment by terracoda** [comment]
@zepumph, I'll move the Firefox focus problem to a separate issue.

**Comment by jessegreenberg** [comment]
@zepumph this sounds great in NVDA with Firefox! Just documenting what I hear in molarity to compare with https://github.com/phetsims/sun/issues/314#issuecomment-433737115:
On focus: "Solute: Nickel roman 2 chloride, button, sub-menu"
On open: "Solute list, bullet, Nickel roman 2 chloride, seven of nine"
On selection: "Solute: Potassium dichromate, button, sub-menu"

I wasn't able to reproduce https://github.com/phetsims/sun/issues/314#issuecomment-433739180 in Windows unfortunately, may be OS specific.

Regarding aria-activedescendant, my understanding is that it is meant to be used when changing focus to child elements of a widget is not possible. In this case, DOM focus would stay on the listbox and the arrow keys would only update aria-activedescendant to tell the AT what has focus conceptually. Since focus is not on the element with aria-activedescendant, I don't think it is currently being used by the AT.

So I think we should either use aria-activedescendant OR change focus, but not both. I don't know if one is more correct, @terracoda do you have a preference?

**Comment by terracoda** [comment]
@jessegreenberg, I will review `aria-activedescendent`.

**Comment by terracoda** [comment]
@jessegreenberg, you are correct, `aria-activedescendant` is an alternative way to manage focus on composite widgets - elements that have logical descendants like comboboxes, menus, tablists, etc.

You either manage focus directly in the DOM, or with `aria-activedescendant`, but not both. The main thing is to manage focus effectively, so users know where they are and what they have selected.

**Comment by terracoda** [comment]
I can share the accessible combobox with an ARIA expert as soon as we think it is ready.

**Comment by emily-phet** [comment]
@zepumph Can you summarize where work on this is? Just curious, as there is description work officially started for this sim, so early 2019 we'll want to move forward with this one.

**Comment by terracoda** [comment]
@emily-phet, I think @zepumph's summary is in https://github.com/phetsims/sun/issues/314#issuecomment-433661093

I think the combobox itself is ready to share with an ARIA expert. Here are some VoiceOver screenshots that show how it sounds with VoiceOver & Safari:
**1. I have opened the list of solutes with the pop-up button**
- and moved my selection from item 1 to item 7, but not yet selected it
- notice the name of each solute and its count (1 of 9, 2 of 9) are read out nicely
<img width="971" alt="new-item-has-focus" src="https://user-images.githubusercontent.com/1902116/49280127-9dcdb080-f489-11e8-8a0a-4d8ec456d000.png">

**2. I selected the new solute (Space or Enter key)** 
- the pop-up closes
- the selected solute is read out with the addition of "Solute"
- Note: the label, "Solute", is read out on initial focus of the pop-up button and after the solute is selected and focus is returned to the pop-up button
<img width="973" alt="molarity-new-item-selected" src="https://user-images.githubusercontent.com/1902116/49280129-9f977400-f489-11e8-8843-08fe344dad8d.png">

**Comment by zepumph** [comment]
Yes I have not worked on anything since https://github.com/phetsims/sun/issues/314#issuecomment-433661093. There is still work to be done, but it fell off my plate because many of the comments after my work were more design discussion rather than clear direction in how to implement a few of the question I have. I agree that an aria expert would be good here. I also think it would be good to discuss at a shorts meeting just to make sure we are all on the same page.

**Comment by terracoda** [comment]
I can share with Bryan Garaventa. The only issue with the link as the only thing within the app that is access is the combobox. Nothing else has labels.

I'll let him know that and instruct him to only investigate the combobox.

**Comment by terracoda** [comment]
Just linking to the original a11y-research issue https://github.com/phetsims/a11y-research/issues/64, in case we need any of those resources to improve our combobox.

**Comment by accdc** [comment]
Hi,
My apologies again for taking so long in getting back.

I've tested the page at
https://phet-dev.colorado.edu/html/molarity/1.5.0-dev.1/phet/molarity_en_phet.html?a11y
And there are indeed some areas that are of concern.

It looks like ARIA is being used in combination with HTML5 interactive widgets such as role=slider plus the HTML5 range widget, and though these map to the same type of control in the accessibility tree, the combination of ARIA roles with interactive HTML widgets causes AT behavioral conflicts that cannot be predicted. As such, ARIA roles such as these should only be used when not implementing native controls for similar purposes. E.G. Adding role=checkbox to input+type=checkbox, or role=listbox or role=combobox on native select elements, and so on will lead to significant accessibility issues down the road. It is for the same reason that HTML5 widgets such as 'details' cannot be made accessible in IE11 using ARIA without breaking its native functionality at the same time in supporting browsers like FF and Chrome.

The listbox and menu popups look great though, and work exactly as they should.

The only things I'd recommend regarding the popup listbox and menu, is to also add key handlers to allow for Home/End key presses to move to the beginning and end of the listbox when focus is within it, and to allow for the pressing of letters or numbers in both listbox and menu constructs to jump to the first options that match what is typed. This additional functionality is very helpful when dealing with large lists or menus, especially for non-sighted screen reader users and for those who rely on speech navigation software such as Dragon. 

I tried to locate a combobox for testing too, but I wasn't sure if this was part of the same page.

All the best,
Bryan

**Comment by terracoda** [comment]
@accdc, thanks for your feedback!

You are correct, there is no actual `combobox` role being used. We are using a popped up `listbox` to represent the visual widget which we all refer to as a combobox.

As for the other issue about mixing native html5 and aria widget roles, I agree it is not the best practice to do so. That said, we tested the combination of range and slider on all our supported platforms and it was the combination of the two that unexpectedly provided the best experience in all cases.

Again, though, the combination of FireFox and NVDA now actually say, "slider" when using `input type="range"`, so we have updated our design pattern to not use the duplicate aria widget `role="slider"` on these range-type interactive objects.

I will discuss these added shortcuts with our developer, @zepumph, to see if we can make them into nice options when we have a long-ish comboboxes. Our comboboxes generally are not very long. In the case of the Solute combobox in this example, I thought we were on the edge of the category of a short list of options.

Again, @accdc thanks for checking out our implementation!

**Comment by pixelzoom** [comment]
I'm having to make what should be a straightforward change to ComboBox in https://github.com/phetsims/sun/issues/428#issuecomment-451554852.   And that has become very complicated because of how a11y was added to ComboBox.  Lack of documentation, methods added inside of constructors, TODO's with no associated GitHub issue,...   Highly recommend to (1) put a11y changes through the same code review process that's used for all other changes, (2) don't push things to master that are partially finished unless you're going to finish them in the short-term; otherwise use a branch. @ariel-phet FYI.

**Comment by terracoda** [comment]
@pixelzoom, the combobox design pattern is documented here:
https://github.com/phetsims/sun/blob/master/doc/ComboBox.md

Is that helpful?

**Comment by pixelzoom** [comment]
Thanks @terracoda.  But it's the way that the design was implemented that is the problem.

**Comment by terracoda** [comment]
Ok, @pixelzoom, I'll let @zepumph comment as he may not be fully finished the implementation.

**Comment by zepumph** [comment]
This implementation got stalled while waiting for feedback and then Friction took over in priority. I agree that the code was not in an acceptable state. @pixelzoom let me know when your work has completed and I will go in and clean up accessibility related features now that we have a good understanding of what we want the final a11y interaction to look like.

**Comment by pixelzoom** [comment]
In the previous commit, I added a bunch of TODOs related to this issue.

**Comment by pixelzoom** [comment]
In #428, items in the combo box (ItemNode instances) are now shared by the list and button.  A couple of things:

* I'm now unable to select items in the list.

* Since ItemNodes are now shared, we can't transform them, highlight them, etc. So they are wrapped by the list and button.  The a11y implementation is currently setting focus on the ItemNode instances, via ItemNode `a11yFocusButton`, and I'm wondering if (a) that is still the correct thing to do, and (b) whether it will cause problems, and (c) what the focus issues are (in general) for Nodes with multiple parents.

**Comment by pixelzoom** [comment]
@zepumph I'm going to pause work on ComboBox so that a11y can be fixed (I think it may currently be broken) and cleaned up.   You should know that the internal structure of ComboBox will be changing dramatically as the result of https://github.com/phetsims/sun/issues/430 -- I'll be formalizing the "list" and "item" by creating inner classes for them.

**Comment by zepumph** [comment]
>what the focus issues are (in general) for Nodes with multiple parents.

In general we don't support focusing any node that has multiple parents, instead I think we would just try to focus a container for it, I'll do some investigation.

**Comment by terracoda** [comment]
@pixelzoom, when you examine the example HTML in the design pattern do you see how this fits into the design pattern of a visual ComboBox?

We have a button that has a dynamic label, so that when focus is placed on the Solute button, a blind user knows what Solute is currently selected before activating the actual Solute button.

Once the user activates the button, a navigable list of solutes is popped up, and one of those solute items has to have focus (the one that was already selected). A user can then navigate the solute items and make their selection, or close the parent popped up list if they do not want to make a new selection. When the list of solutes is closed, the user's focus is returned to the Solute button which confirms which solute is selected.

All of that information needs to be communicated and managed visually and non-visually.

To complicate implementation, we don't even use the ARIA widget `role="combobox"` because that role is not yet well-suppoted by Assistive Technologies. We have to use a popped-up list instead and this seems to work well to communicate the experience of combobox and to make it accessible to more users at the same time.

Designing for A11y is tricky and evolving. I am happy to see your interest in helping and guiding improving how we do things, so it can be as robust as possible from all perspectives.

**Comment by pixelzoom** [comment]
Thanks for the overview @terracoda, much appreciated.  Again, I don't have any problem with the overall design or philosophy of how ComboBox is instrumented.  It's the implementation that is overly complicated and lacking, and it doesn't need to be that way to support the design.

**Comment by terracoda** [comment]
@pixelzoom, great! I am sure @zepumph is going to fix it up nicely for us!

**Comment by pixelzoom** [comment]
I'm done with changes to ComboBox see #430.  I preserved all TODO items related to this issue; search for "TODO sun#314".

Following that work, I confirmed that a11y still works via the following test:

* Ran Beer's Law Lab with `?a11y`. Confirmed that keyboard navigation of ComboBox works. I could navigate to the button (tab), pop up the listbox (spacebar), move between items in the list (arrow keys), and select an item in the list (spacebar).

I did not succeed in factoring out `ComboBoxListBox`, the subject of https://github.com/phetsims/sun/issues/445. Highly recommended to do that before further a11y work on ComboBox.

Over to @zepumph.

**Comment by pixelzoom** [comment]
@zepumph and I spent many hours today mopping up TODO items related to this issue.  We used ComboBox.md as the ground through, but that document should be reviewed and possibly updated, see https://github.com/phetsims/sun/issues/465.

Since all TODO comments related to this issue have been addressed or moved to new items....  Closing.
